lib/operations/db_ops.js

```diff
@@ -0,0 +1,980 @@
+
+  // Ensure we have a callback
+  if (finalOptions.writeConcern && typeof callback !== 'function') {
+    throw MongoError.create({
+      message: 'Cannot use a writeConcern without a provided callback',
+      driver: true
+    });
+  }
+
+  // Did the user destroy the topology
+  if (db.serverConfig && db.serverConfig.isDestroyed())
+    return callback(new MongoError('topology was destroyed'));
+
+  // Attempt to run using createIndexes command
+  _createIndexUsingCreateIndexes(db, name, fieldOrSpec, options, (err, result) => {
+    if (err == null) return handleCallback(callback, err, result);
+
+    /**
+     * The following errors mean that the server recognized `createIndex` as a command so we don't need to fallback to an insert:
+     * 67 = 'CannotCreateIndex' (malformed index options)
+     * 85 = 'IndexOptionsConflict' (index already exists with different options)
+     * 86 = 'IndexKeySpecsConflict' (index already exists with the same name)
+     * 11000 = 'DuplicateKey' (couldn't build unique index because of dupes)
+     * 11600 = 'InterruptedAtShutdown' (interrupted at shutdown)
+     */
+    if (
+      err.code === 67 ||
+      err.code === 11000 ||
+      err.code === 85 ||
+      err.code === 86 ||
+      err.code === 11600
+    ) {
+      return handleCallback(callback, err, result);
+    }
+
+    // Create command
+    const doc = _createCreateIndexCommand(db, name, fieldOrSpec, options);
+    // Set no key checking
+    finalOptions.checkKeys = false;
+    // Insert document
+    db.s.topology.insert(
+      `${db.s.databaseName}.${Db.SYSTEM_INDEX_COLLECTION}`,
+      doc,
+      finalOptions,
+      (err, result) => {
+        if (callback == null) return;
+        if (err) return handleCallback(callback, err);
+        if (result == null) return handleCallback(callback, null, null);
+        if (result.result.writeErrors)
+          return handleCallback(callback, MongoError.create(result.result.writeErrors[0]), null);
+        handleCallback(callback, null, doc.name);
+      }
+    );
+  });
+}
+
+// Add listeners to topology
+function createListener(db, e, object) {
+  function listener(err) {
+    if (object.listeners(e).length > 0) {
+      object.emit(e, err, db);
+
+      // Emit on all associated db's if available
+      for (let i = 0; i < db.s.children.length; i++) {
+        db.s.children[i].emit(e, err, db.s.children[i]);
+      }
+    }
+  }
+  return listener;
+}
+
+/**
+ * Drop a collection from the database, removing it permanently. New accesses will create a new collection.
+ *
+ * @method
+ * @param {Db} db The Db instance on which to drop the collection.
+ * @param {string} name Name of collection to drop
+ * @param {Object} [options] Optional settings. See Db.prototype.dropCollection for a list of options.
+ * @param {Db~resultCallback} [callback] The results callback
+ */
+function dropCollection(db, cmd, options, callback) {
+  return executeCommand(db, cmd, options, (err, result) => {
```

Comment:
I think we need to be very wary of situations like this as well. The operations are all meant to explicitly use callbacks, so we can support callbacks and promises in `executeOperation`. This method is returning `executeCommand` which visually appears to be returning a promise, even though it doesn't since the callback is provided. Still, I think we should make things all look correct, and furthermore try not to mix promises and callbacks as much as possible.